tests/network/vmi_istio.go

```diff
@@ -0,0 +1,304 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	expect "github.com/google/goexpect"
+	k8snetworkplumbingwgv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libvmi"
+)
+
+const (
+	istioInjectSidecarAnnotation = "sidecar.istio.io/inject"
+	istioDeployedEnvVariable     = "KUBEVIRT_DEPLOY_ISTIO"
+	vmiAppSelector               = "istio-vmi-app"
+	svcDeclaredTestPort          = 1500
+	svcUndeclaredTestPort        = 1501
+)
+
+var _ = SIGDescribe("[Serial] Istio", func() {
+	var (
+		err        error
+		vmi        *v1.VirtualMachineInstance
+		virtClient kubecli.KubevirtClient
+	)
+	BeforeEach(func() {
+		if !istioServiceMeshDeployed() {
+			Skip("Istio service mesh is required for service-mesh tests to run")
+		}
+	})
+
+	Context("Virtual Machine with masquerade interface and explicitly specified ports", func() {
+		createJobCheckingVMIReachability := func(serverVMI *v1.VirtualMachineInstance, targetPort int) (*batchv1.Job, error) {
+			By("Starting HTTP Server")
+			tests.StartHTTPServer(vmi, targetPort)
+
+			By("Getting back the VMI IP")
+			vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
+			vmiIP := vmi.Status.Interfaces[0].IP
+
+			By("Running job to send a request to the server")
+			return virtClient.BatchV1().Jobs(tests.NamespaceTestDefault).Create(
+				context.Background(),
+				tests.NewHelloWorldJobHTTP(vmiIP, fmt.Sprintf("%d", targetPort)),
+				metav1.CreateOptions{},
+			)
+		}
+		BeforeEach(func() {
+			tests.BeforeTestCleanup()
+
+			virtClient, err = kubecli.GetKubevirtClient()
+			tests.PanicOnError(err)
+
+			// Enable sidecar injection by setting the namespace label
+			Expect(libnet.AddLabelToNamespace(virtClient, tests.NamespaceTestDefault, tests.IstioInjectNamespaceLabel, "enabled")).ShouldNot(HaveOccurred())
+			defer func() {
```

Comment:
> EDIT: oh, you create the vm in the BeforeEach so I guess the label is not needed anymore after the vm is created. Anyway, I find it confusing.

Yes, I only add the label when creating the VMI that's to be tested. 

 > EDIT2: You've added the label removal to the general cleanup method. Why this specific cleanup is needed as well?
 
It shouldn't be needed. I added the removal to be extra sure that the label doesn't leak to other tests somehow.
But it shouldn't since I'm deferring the removal